File: Design-Patterns-Using-Swift/Bridge/BridgePatternShape.swift
// Abstraction protocol
protocol Shape {
    func draw()
}

// Implementor protocol
protocol Color {
    func fill() -> String
}

// Rectangle class
class Rectangle: Shape {
    private var imp: Color

    init(imp: Color) {
        self.imp = imp
    }

    func draw() {
        print("Drawing Rectangle with color \(imp.fill())")
    }
}

// Circle class
class Circle: Shape {
    private var imp: Color

    init(imp: Color) {
        self.imp = imp
    }

    func draw() {
        print("Drawing Circle with color \(imp.fill())")
    }
}

// Red class
class Red: Color {
    func fill() -> String {
        return "Red"
    }
}

// Green class
class Green: Color {
    func fill() -> String {
        return "Green"
    }
}

// Blue class
class Blue: Color {
    func fill() -> String {
        return "Blue"
    }
}

// Client code
let c1 = Red()
let abstraction = Circle(imp: c1)
abstraction.draw()

let c2 = Green()
let abstraction2 = Rectangle(imp: c2)
abstraction2.draw()

/*
Drawing Circle with color Red
Drawing Rectangle with color Green
*/

File: Design-Patterns-Using-Ruby/Bridge/BridgePatternShape.rb
# Abstraction interface
class Shape
    def initialize(imp)
        @imp = imp
    end
    
    def draw
        puts "Drawing #{self.class} with color #{@imp.fill}"
    end
end

# Implementor interface
class Color
    def fill
        raise NotImplementedError, 'Subclasses must implement this method'
    end
end

# Rectangle class
class Rectangle < Shape
end

# Circle class
class Circle < Shape
end

# Red class
class Red < Color
    def fill
        'Red'
    end
end

# Green class
class Green < Color
    def fill
        'Green'
    end
end

# Blue class
class Blue < Color
    def fill
        'Blue'
    end
end

# Client code
c1 = Red.new
abstraction = Circle.new(c1)
abstraction.draw

c2 = Green.new
abstraction2 = Rectangle.new(c2)
abstraction2.draw

=begin 
Drawing Circle with color Red
Drawing Rectangle with color Green 
=end